pkg/ctrl/controller.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ctrl
+
+// Reconcile enqueues ObjectKeys for a Reconcile to Reconcile.
+type Controller struct {
+	// Reconcile Reconciles observed and desired object state within a cluster.
+	// Defaults to the DefaultReconcileFn
+	Reconcile Reconcile
+
+	// Name is the name that will appear in tracing, logging and monitoring metrics.
+	// Defaults to a generated value
+	Name string
+
+	// ControllerParallelism is the number of concurrent Reconciles
+	// Defaults to 1
+	ControllerParallelism int
+
+	// Stop is used to shutdown the Reconcile
+	// Defaults to a new channel
+	Stop <-chan struct{}
+
+	// BeforeReconcile is called before Reconciling an ObjectKey.
+	// Defaults to no-op
+	BeforeReconcile func(ObjectKey)
+
+	// AfterReconcile is called after Reconciling an ObjectKey
+	// Defaults to no-op
+	AfterReconcile func(ObjectKey, err error)
+}
+
+// WatchResource enqueues events from resource events.
+func (*Controller) WatchResource(Watchable, EventHandler) {
+	// TODO: Write this
+}
+
+// WatchChannel enqueues events from the channel
+func (*Controller) WatchChannel(chan interface{}, EventHandler) {
+	// TODO: Write this
+}
+
+// Start starts the controller watching for events and reconciling
```

Comment:
That should say if this is blocking or not? Looks like your controller manager is blocking, but it's also assuming these are non-blocking, which is inconsistent? Make them all blocking and expect people to use goroutines, and use the channel to stop them?

Reply:
Done.  Yes making them blocking.